### ISSUE TITLE
Files being deleted before multiple merging is complete

### ISSUE THREAD

**Issue by atom-smasher** [ytdl-org/youtube-dl#29795]
## Checklist

- [ ] I'm reporting a broken site support issue
- [x] I've verified that I'm running youtube-dl version **2021.06.06**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar bug reports including closed ones
- [x] I've read bugs section in FAQ

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'--ignore-config', u'-f', u'webm[height<1090]+m4a,webm[height<400]+m4a', u'https://www.youtube.com/watch?v=XS9o31UBBTI', u'-o', u'%(title)s-%(id)s-%(height)sp.%(ext)s']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2021.06.06
[debug] Python version 2.7.18 (CPython) - Linux-5.4.0-80-generic-x86_64-with-Ubuntu-20.04-focal
[debug] exe versions: ffmpeg 4.2.4, ffprobe 4.2.4, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[youtube] XS9o31UBBTI: Downloading webpage
[info] XS9o31UBBTI: downloading video in 2 formats
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[debug] Invoking downloader on u'https://r1---sn-uo1-fabl.googlevideo.com/videoplayback?expire=1629018959&ei=74YYYcLuNKDd4-EPwd-g6AM&ip=122.58.109.177&id=o-AF0MTgPAF02duqzD2xDdyW3awMdFXHHb05cXfrRVLsD4&itag=244&aitags=133%2C134%2C135%2C136%2C137%2C160%2C242%2C243%2C244%2C247%2C248%2C278%2C394%2C395%2C396%2C397%2C398%2C399&source=youtube&requiressl=yes&mh=a-&mm=31%2C29&mn=sn-uo1-fabl%2Csn-ntqe6nel&ms=au%2Crdu&mv=m&mvi=1&pcm2cms=yes&pl=21&initcwndbps=1523750&vprv=1&mime=video%2Fwebm&ns=eYO7PlI5zD3pe6hH_5XBl0cG&gir=yes&clen=799782&dur=14.333&lmt=1625441250568891&mt=1628997169&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5535432&n=TBF08-LPJLYdGtuamz5&sparams=expire%2Cei%2Cip%2Cid%2Caitags%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIhAPlsXiojhdURDW9clCi4o-iTz9U8k8sVdP9xlOmSiNtzAiAuK794lnEHbMWcn08VMDJ7kQsCujA3wJJOOifZgSKilA%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhALChyse437cG86fbostkS2PCKrM6H4aaRkObIQEofl9jAiBXSBSxzgnTzxZ-Ef2-oCmA1-WvUSnWo-5RgsEG3r5eKw%3D%3D'
[download] Destination: A Penny Short #shorts-XS9o31UBBTI-854p.f244.webm
[download] 100% of 781.04KiB in 00:00
[debug] Invoking downloader on u'https://r1---sn-uo1-fabl.googlevideo.com/videoplayback?expire=1629018959&ei=74YYYcLuNKDd4-EPwd-g6AM&ip=122.58.109.177&id=o-AF0MTgPAF02duqzD2xDdyW3awMdFXHHb05cXfrRVLsD4&itag=140&source=youtube&requiressl=yes&mh=a-&mm=31%2C29&mn=sn-uo1-fabl%2Csn-ntqe6nel&ms=au%2Crdu&mv=m&mvi=1&pcm2cms=yes&pl=21&initcwndbps=1523750&vprv=1&mime=audio%2Fmp4&ns=eYO7PlI5zD3pe6hH_5XBl0cG&gir=yes&clen=233961&dur=14.396&lmt=1625440437183527&mt=1628997169&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5531432&n=TBF08-LPJLYdGtuamz5&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRgIhAJXM-E_L8gfJtvX7AFsyhsaBwhYJZcgNnD4P7LAYYAMxAiEAlojvLxXPRWw5mlFP8M037Dyz0G8V7UnD0KA-cKfG_50%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhALChyse437cG86fbostkS2PCKrM6H4aaRkObIQEofl9jAiBXSBSxzgnTzxZ-Ef2-oCmA1-WvUSnWo-5RgsEG3r5eKw%3D%3D'
[download] Destination: A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a
[download] 100% of 228.48KiB in 00:00
[ffmpeg] Merging formats into "A Penny Short #shorts-XS9o31UBBTI-854p.mkv"
[debug] ffmpeg command line: ffmpeg -y -loglevel 'repeat+info' -i 'file:A Penny Short #shorts-XS9o31UBBTI-854p.f244.webm' -i 'file:A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a' -c copy -map '0:v:0' -map '1:a:0' 'file:A Penny Short #shorts-XS9o31UBBTI-854p.temp.mkv'
Deleting original file A Penny Short #shorts-XS9o31UBBTI-854p.f244.webm (pass -k to keep)
Deleting original file A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a (pass -k to keep)
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[debug] Invoking downloader on u'https://r1---sn-uo1-fabl.googlevideo.com/videoplayback?expire=1629018959&ei=74YYYcLuNKDd4-EPwd-g6AM&ip=122.58.109.177&id=o-AF0MTgPAF02duqzD2xDdyW3awMdFXHHb05cXfrRVLsD4&itag=278&aitags=133%2C134%2C135%2C136%2C137%2C160%2C242%2C243%2C244%2C247%2C248%2C278%2C394%2C395%2C396%2C397%2C398%2C399&source=youtube&requiressl=yes&mh=a-&mm=31%2C29&mn=sn-uo1-fabl%2Csn-ntqe6nel&ms=au%2Crdu&mv=m&mvi=1&pcm2cms=yes&pl=21&initcwndbps=1523750&vprv=1&mime=video%2Fwebm&ns=eYO7PlI5zD3pe6hH_5XBl0cG&gir=yes&clen=172147&dur=14.333&lmt=1625441250574347&mt=1628997169&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5535432&n=TBF08-LPJLYdGtuamz5&sparams=expire%2Cei%2Cip%2Cid%2Caitags%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRgIhANMa3kvvamqmzN3HpOHxEatQrOTQAVxzbDiM9qIOi9hvAiEAmGSZ0uBjg33sZ7NYWAolwDt33aPaNcNIa9JRrYoyEmA%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhALChyse437cG86fbostkS2PCKrM6H4aaRkObIQEofl9jAiBXSBSxzgnTzxZ-Ef2-oCmA1-WvUSnWo-5RgsEG3r5eKw%3D%3D'
[download] Destination: A Penny Short #shorts-XS9o31UBBTI-256p.f278.webm
[download] 100% of 168.11KiB in 00:00
[debug] Invoking downloader on u'https://r1---sn-uo1-fabl.googlevideo.com/videoplayback?expire=1629018959&ei=74YYYcLuNKDd4-EPwd-g6AM&ip=122.58.109.177&id=o-AF0MTgPAF02duqzD2xDdyW3awMdFXHHb05cXfrRVLsD4&itag=140&source=youtube&requiressl=yes&mh=a-&mm=31%2C29&mn=sn-uo1-fabl%2Csn-ntqe6nel&ms=au%2Crdu&mv=m&mvi=1&pcm2cms=yes&pl=21&initcwndbps=1523750&vprv=1&mime=audio%2Fmp4&ns=eYO7PlI5zD3pe6hH_5XBl0cG&gir=yes&clen=233961&dur=14.396&lmt=1625440437183527&mt=1628997169&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5531432&n=TBF08-LPJLYdGtuamz5&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRgIhAJXM-E_L8gfJtvX7AFsyhsaBwhYJZcgNnD4P7LAYYAMxAiEAlojvLxXPRWw5mlFP8M037Dyz0G8V7UnD0KA-cKfG_50%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhALChyse437cG86fbostkS2PCKrM6H4aaRkObIQEofl9jAiBXSBSxzgnTzxZ-Ef2-oCmA1-WvUSnWo-5RgsEG3r5eKw%3D%3D'
[download] Destination: A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a
[download] 100% of 228.48KiB in 00:00
[ffmpeg] Merging formats into "A Penny Short #shorts-XS9o31UBBTI-256p.mkv"
[debug] ffmpeg command line: ffmpeg -y -loglevel 'repeat+info' -i 'file:A Penny Short #shorts-XS9o31UBBTI-256p.f278.webm' -i 'file:A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a' -c copy -map '0:v:0' -map '1:a:0' 'file:A Penny Short #shorts-XS9o31UBBTI-256p.temp.mkv'
Deleting original file A Penny Short #shorts-XS9o31UBBTI-256p.f278.webm (pass -k to keep)
Deleting original file A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a (pass -k to keep)

```

## Description

Expected behaviour: Using a comma separated list to download files with different qualities (eg `-f "webm[height<1090]+m4a","webm[height<400]+m4a"`) I expect the downloaded files to be kept until after all merge operations are complete.

Actual behaviour: When using a comma separated list to download files with different qualities (eg `-f "webm[height<1090]+m4a","webm[height<400]+m4a"`) the files are deleted after *each* merge. This behaviour requires the same file (eg `video-xyz.f140.m4a`) to be downloaded multiple times.

As the above example demonstrates, I should end up with two copies of the same video, both with "best" m4a audio, and that part works fine. Running that as a single command, I expect that the audio file (eg `video-xyz.f140.m4a`) would only be downloaded once, then deleted after all merges are complete.

Different combinations of preferences for downloading the same video in different video/audio qualities will result in different combinations of files that are downloaded multiple times. In this particular example, it's the m4a audio file that's downloaded multiple times.

Workaround: Using the `-k` or `--keep-video` option, the audio file is only downloaded once, but then clutter needs to be manually cleaned up.

Desired behaviour: Clean-up of downloaded files should happen before exit, not after each merge.

**Comment by dirkf** [comment]
A format countdown could be added, delaying the temporary file deletion until the countdown reaches 0.
```
# youtube-dl -v --ignore-config -f 'webm[height<1090]+m4a,webm[height<400]+m4a' 'https://www.youtube.com/watch?v=XS9o31UBBTI' -o '%(title)s-%(id)s-%(height)sp.%(ext)s'
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'--ignore-config', u'-f', u'webm[height<1090]+m4a,webm[height<400]+m4a', u'https://www.youtube.com/watch?v=XS9o31UBBTI', u'-o', u'%(title)s-%(id)s-%(height)sp.%(ext)s']
[debug] Encodings: locale ASCII, fs ASCII, out ASCII, pref ASCII
[debug] youtube-dl version 2021.06.06.1
[debug] Python version 2.7.1 (CPython) - Linux-2.6.18-7.1-7405b0-smp-with-libc0
[debug] exe versions: ffmpeg 4.1, ffprobe 4.1
[debug] Proxy map: {}
[youtube] XS9o31UBBTI: Downloading webpage
[info] XS9o31UBBTI: downloading video in 2 formats
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[debug] Invoking downloader on u'https://r1---sn-cu-aigsl.googlevideo.com/videoplayback?expire=1629134990&ei=LkwaYZ7lNNiJmLAPieaWGA&ip=51.6.64.237&id=o-AE_sIwGyHGsEl41j_Ld0JEuskGGamTXauacfHm1WBIy3&itag=244&aitags=133%2C134%2C135%2C136%2C137%2C160%2C242%2C243%2C244%2C247%2C248%2C278%2C394%2C395%2C396%2C397%2C398%2C399&source=youtube&requiressl=yes&mh=a-&mm=31%2C29mn=sn-cu-aigsl%2Csn-cu-c9ie&ms=au%2Crdu&mv=m&mvi=1&pl=27&initcwndbps=1257500&vprv=1&mime=video%2Fwebm&ns=9iZhN3wx72U7c32O16hRl_EG&gir=yes&clen=799782&dur=14.333&lmt=1625441250568891&mt=1629113082&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5535432&n=WT3XroqDS_rXT3j_Rvv&sparams=expire%2Cei%2Cip%2Cid%2Caitags%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIhANRfJZg5n-k-MBJG9BJmqLnrNvtZbwOE7CjiDK_LUv5wAiACNUOQtZp9Y4PJ9nxAa_g4nPiSp_jKiZKwe2U_EDu1Pw%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgAgJxegJizlyiS0lb9CHzutTngvOhT_JcUXKC7B4KHN4CIQDRh9EwYlTeUbzeU4wHEb81hp_7nYMxD1fzwK9uzA8jGA%3D%3D'
[download] Destination: A Penny Short #shorts-XS9o31UBBTI-854p.f244.webm
[download] 100% of 781.04KiB in 00:00
[debug] Invoking downloader on u'https://r1---sn-cu-aigsl.googlevideo.com/videoplayback?expire=1629134990&ei=LkwaYZ7lNNiJmLAPieaWGA&ip=51.6.64.237&id=o-AE_sIwGyHGsEl41j_Ld0JEuskGGamTXauacfHm1WBIy3&itag=140&source=youtube&requiressl=yes&mh=a-&mm=31%2C29&mn=sn-cu-aigsl%2Csn-cu-c9ie&ms=au%2Crdu&mv=m&mvi=1&pl=27&initcwndbps=1257500&vprv=1&mime=audio%2Fmp4&ns=9iZhN3wx72U7c32O16hRl_EG&gir=yes&clen=233961&dur=14.396&lmt=1625440437183527&mt=1629113082&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5531432&n=WT3XroqDS_rXT3j_Rvv&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgN34ZVREsNgxEmUNJLFV8HgtLKvQnV-iZ19Q_MA8v87QCIBCbviIQTfCGI_2pMqD3EvBqWEFZLc4CzJMEAvACWUQF&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgAgJxegJizlyiS0lb9CHzutTngvOhT_JcUXKC7B4KHN4CIQDRh9EwYlTeUbzeU4wHEb81hp_7nYMxD1fzwK9uzA8jGA%3D%3D'
[download] Destination: A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a
[download] 100% of 228.48KiB in 00:00
[ffmpeg] Merging formats into "A Penny Short #shorts-XS9o31UBBTI-854p.mkv"
[debug] ffmpeg command line: ffmpeg -y -loglevel 'repeat+info' -i 'file:A Penny Short #shorts-XS9o31UBBTI-854p.f244.webm' -i 'file:A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a' -c copy -map '0:v:0' -map '1:a:0' 'file:A Penny Short #shorts-XS9o31UBBTI-854p.temp.mkv'
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[debug] Invoking downloader on u'https://r1---sn-cu-aigsl.googlevideo.com/videoplayback?expire=1629134990&ei=LkwaYZ7lNNiJmLAPieaWGA&ip=51.6.64.237&id=o-AE_sIwGyHGsEl41j_Ld0JEuskGGamTXauacfHm1WBIy3&itag=278&aitags=133%2C134%2C135%2C136%2C137%2C160%2C242%2C243%2C244%2C247%2C248%2C278%2C394%2C395%2C396%2C397%2C398%2C399&source=youtube&requiressl=yes&mh=a-&mm=31%2C29&mn=sn-cu-aigsl%2Csn-cu-c9ie&ms=au%2Crdu&mv=m&mvi=1&pl=27&initcwndbps=1257500&vprv=1&mime=video%2Fwebm&ns=9iZhN3wx72U7c32O16hRl_EG&gir=yes&clen=172147&dur=14.333&lmt=1625441250574347&mt=1629113082&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5535432&n=WT3XroqDS_rXT3j_Rvv&sparams=expire%2Cei%2Cip%2Cid%2Caitags%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRgIhAIN_ibAamdSuXwxKw_jRaMkKvbMnXCzzzwujNuw87EzoAiEA5d5mO90fUkOzAYI9xcv8S7orxgWRIbmsVQIx5YQnsAw%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgAgJxegJizlyiS0lb9CHzutTngvOhT_JcUXKC7B4KHN4CIQDRh9EwYlTeUbzeU4wHEb81hp_7nYMxD1fzwK9uzA8jGA%3D%3D'
[download] Destination: A Penny Short #shorts-XS9o31UBBTI-256p.f278.webm
[download] 100% of 168.11KiB in 00:00
[debug] Invoking downloader on u'https://r1---sn-cu-aigsl.googlevideo.com/videoplayback?expire=1629134990&ei=LkwaYZ7lNNiJmLAPieaWGA&ip=51.6.64.237&id=o-AE_sIwGyHGsEl41j_Ld0JEuskGGamTXauacfHm1WBIy3&itag=140&source=youtube&requiressl=yes&mh=a-&mm=31%2C29&mn=sn-cu-aigsl%2Csn-cu-c9ie&ms=au%2Crdu&mv=m&mvi=1&pl=27&initcwndbps=1257500&vprv=1&mime=audio%2Fmp4&ns=9iZhN3wx72U7c32O16hRl_EG&gir=yes&clen=233961&dur=14.396&lmt=1625440437183527&mt=1629113082&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=WEB&txp=5531432&n=WT3XroqDS_rXT3j_Rvv&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgN34ZVREsNgxEmUNJLFV8HgtLKvQnV-iZ19Q_MA8v87QCIBCbviIQTfCGI_2pMqD3EvBqWEFZLc4CzJMEAvACWUQF&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgAgJxegJizlyiS0lb9CHzutTngvOhT_JcUXKC7B4KHN4CIQDRh9EwYlTeUbzeU4wHEb81hp_7nYMxD1fzwK9uzA8jGA%3D%3D'
[download] A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a has already been downloaded
[download] 100% of 228.48KiB
[ffmpeg] Merging formats into "A Penny Short #shorts-XS9o31UBBTI-256p.mkv"
[debug] ffmpeg command line: ffmpeg -y -loglevel 'repeat+info' -i 'file:A Penny Short #shorts-XS9o31UBBTI-256p.f278.webm' -i 'file:A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a' -c copy -map '0:v:0' -map '1:a:0' 'file:A Penny Short #shorts-XS9o31UBBTI-256p.temp.mkv'
Deleting original file A Penny Short #shorts-XS9o31UBBTI-256p.f278.webm (pass -k to keep)
Deleting original file A Penny Short #shorts-XS9o31UBBTI-NAp.f140.m4a (pass -k to keep)
#
```
On the other hand, delaying the deletion might leave too little room for the later operations, or there may be a caching proxy between yt-dl and the site, which would make repeated downloads less of an issue. There would have to be some way to control this behaviour. IMO this change would add complexity for little benefit, compared with just using `--keep-video`.
